src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -390,6 +390,82 @@ def test_all_endpoints_have_valid_composition(self):
     def test_topology_is_valid(self):
         asserts.skip("TODO: Make a test that verifies each endpoint only lists direct descendants, except Root Node and Aggregator endpoints that list all their descendants")
 
+    def test_TC_PS_3_1(self):
+        success = True
+        self.print_step(1, "Wildcard read of device - already done")
+
+        self.print_step(2, "Verify that all endpoints listed in the EndpointList are valid")
+        attribute_id = Clusters.PowerSource.Attributes.EndpointList.attribute_id
+        cluster_id = Clusters.PowerSource.id
+        attribute_string = self.cluster_mapper.get_attribute_string(cluster_id, attribute_id)
+        for endpoint_id, endpoint in self.endpoints.items():
+            if Clusters.PowerSource not in endpoint:
+                continue
+            if Clusters.PowerSource.Attributes.EndpointList not in endpoint[Clusters.PowerSource]:
+                location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                self.record_error(self.get_test_name(), location=location,
+                                  problem=f'Did not find {attribute_string} on {location.as_cluster_string(self.cluster_mapper)}', spec_location="EndpointList Attribute")
+                success = False
+            endpoint_list = endpoint[Clusters.PowerSource][Clusters.PowerSource.Attributes.EndpointList]
```

Comment:
The above verifies if EndpointList is in the list. Should have there been a continue above? otherwise if the condition above fails, this gets an invalid item from the dictionary.

I wonder if it would be better to create some methods on the class that yield values when they exist and report an error when they fail. This method is super long and seems somewhat hard to review.

Reply:
I see you check for endpoint_list ... so I guess None/empty list would still work in the code, however I wonder if we can just bypass that by continue.

Reply:
continue is fine - added.